### PR TITLE
DRILL-7272: Drill Metastore Read / Write API and Drill Iceberg Metastore implementation

### PR DESCRIPTION
**Pull request by arina-ielchiieva**
1. Drill Metastore Read / Write API.
2. Drill Iceberg Metastore implementation in iceberg-metastore module.
3. Patches Guava Preconditions class for Apache Iceberg.
4. General refactoring.
5. Unit tests.
6. Documentation.

Jira - [DRILL-7272](https://issues.apache.org/jira/browse/DRILL-7272).

### CODE REVIEW

**Comment by arina-ielchiieva** [comment]
@vvysotskyi please review.

File: metastore/metastore-api/src/main/java/org/apache/drill/metastore/MetastoreRegistry.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+   * For example, if Iceberg module config defines relative path based on Drill Zk root:
+   * drill.metastore.iceberg.location.relative_path: ${drill.exec.zk.root}"/metastore/iceberg",
+   * and Drill main config defines drill.exec.zk.root as "drill",
+   * resulting Iceberg table relative path will be drill/metastore/iceberg.
+   *
+   * @param config main Drill config
+   * @return metastore config
+   */
+  private DrillConfig createMetastoreConfig(DrillConfig config) {
+    return DrillConfig.create(null, null,true, new MetastoreConfigFileInfo(), config.root());
```
Comment (vvysotskyi) [inline metastore/metastore-api/src/main/java/org/apache/drill/metastore/MetastoreRegistry.java]:
```suggestion
    return DrillConfig.create(null, null, true, new MetastoreConfigFileInfo(), config.root());
```

File: metastore/metastore-api/src/main/java/org/apache/drill/metastore/MetastoreTableInfo.java [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+  public long metastoreVersion() {
+    return metastoreVersion;
+  }
+
+  /**
+   * Checks if table metadata has changed or not, based on given exists status and last modified time.
+   * Checks are done based on the following rules and order:
+   * <ul>
+   *   <li>If table did not exist but now does not, return true.</li>
```
Comment (vvysotskyi) [inline metastore/metastore-api/src/main/java/org/apache/drill/metastore/MetastoreTableInfo.java]:
```suggestion
   *   <li>If table did not exist but now does, return true.</li>
```

File: metastore/metastore-api/src/main/java/org/apache/drill/metastore/MetastoreTableInfo.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+   *   <li>If table existed but now does not, return true.</li>
+   *   <li>If both last modified times are null, return false.</li>
+   *   <li>If one last modified time is null and other is not, return true.</li>
+   *   <li>If both last modified times are the same, return true.</li>
```
Comment (vvysotskyi) [inline metastore/metastore-api/src/main/java/org/apache/drill/metastore/MetastoreTableInfo.java]:
```suggestion
   *   <li>If both last modified times are the same, return false.</li>
```

File: metastore/metastore-api/src/main/java/org/apache/drill/metastore/MetastoreTableInfo.java [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+   *   <li>If both last modified times are different, return false.</li>
```
Comment (vvysotskyi) [inline metastore/metastore-api/src/main/java/org/apache/drill/metastore/MetastoreTableInfo.java]:
```suggestion
   *   <li>If both last modified times are different, return true.</li>
```

File: metastore/metastore-api/src/main/java/org/apache/drill/metastore/expressions/SimplePredicate.java [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+      .toString();
+  }
+
+  /**
+   * Indicates {@link FilterExpression.Operator#EQUAL} operator expression:
+   * storagePlugin = 'dfs'.
+   *
+   * @param <T> expression value type
+   */
+  public static class Equal<T> extends SimplePredicate<T>{
```
Comment (vvysotskyi) [inline metastore/metastore-api/src/main/java/org/apache/drill/metastore/expressions/SimplePredicate.java]:
```suggestion
  public static class Equal<T> extends SimplePredicate<T> {
```

File: metastore/metastore-api/src/main/resources/drill-metastore-default.conf [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+drill.metastore: {
+    implementation.class: "org.apache.drill.metastore.iceberg.IcebergMetastore"
+}
```
Comment (vvysotskyi) [inline metastore/metastore-api/src/main/resources/drill-metastore-default.conf]:
```suggestion
}

```

File: metastore/iceberg-metastore/pom.xml [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+        <dependency>
+          <groupId>org.eclipse.jetty</groupId>
+          <artifactId>jetty-servlet</artifactId>
+          <scope>test</scope>
+        </dependency>
+      </dependencies>
+    </profile>
+  </profiles>
+
+</project>
```
Comment (vvysotskyi) [inline metastore/iceberg-metastore/pom.xml]:
```suggestion
</project>

```

**Comment by arina-ielchiieva** [comment]
@vvysotskyi thanks for the code review, addressed code review comments and added Metastore components functionality. Please review.

**Pull request merged**